slsReceiverSoftware/src/DataProcessor.cpp

```diff
@@ -519,62 +579,216 @@ void DataProcessor::PadMissingPackets(sls_receiver_header header, char *data) {
+    if (ctbDigitalDataBytes == 0) {
+        LOG(logWARNING)
+            << "No digital data for call back, yet reorder is set to 1.";
+        return;
+    }
+
+    // make sure data is aligned to 8 bytes before casting to uint64_t
+    AlignedData<uint64_t> aligned_data(nullptr, nullptr);
+    AlignData<uint64_t>(aligned_data, data + nAnalogDataBytes + ctbDbitOffset,
+                        ctbDigitalDataBytes);
+
+    uint64_t *source = aligned_data.ptr;
+
+    const size_t numDigitalSamples = (ctbDigitalDataBytes / sizeof(uint64_t));
+
+    size_t numBytesPerBit =
+        (numDigitalSamples % 8 == 0)
+            ? numDigitalSamples / 8
+            : numDigitalSamples / 8 +
+                  1; // number of bytes per bit in digital data after reordering
+
+    size_t totalNumBytes =
+        numBytesPerBit *
+        64; // number of bytes for digital data after reordering
+
+    std::vector<uint8_t> result(totalNumBytes, 0);
+    uint8_t *dest = &result[0];
 
-    // loop through digital bit enable vector
     int bitoffset = 0;
-    for (auto bi : ctbDbitList) {
-        // where numbits * numDigitalSamples is not a multiple of 8
+    // reorder
+    for (size_t bi = 0; bi < 64; ++bi) {
+
         if (bitoffset != 0) {
             bitoffset = 0;
             ++dest;
         }
 
-        // loop through the frame digital data
-        for (auto *ptr = source; ptr < (source + numDigitalSamples);) {
-            // get selected bit from each 8 bit
-            uint8_t bit = (*ptr++ >> bi) & 1;
-            *dest |= bit << bitoffset;
+        for (auto *ptr = source; ptr < (source + numDigitalSamples); ++ptr) {
+            uint8_t bit = (*ptr >> bi) & 1;
+            *dest |= bit << bitoffset; // most significant bits will be padded
             ++bitoffset;
-            // extract destination in 8 bit batches
             if (bitoffset == 8) {
                 bitoffset = 0;
                 ++dest;
             }
         }
     }
 
+    // move transceiver data to not overwrite and avoid gap in memory
+    if (totalNumBytes != nDigitalDataBytes)
+        memmove(data + nAnalogDataBytes + totalNumBytes * sizeof(uint8_t),
+                data + nAnalogDataBytes + nDigitalDataBytes,
+                nTransceiverDataBytes);
+
     // copy back to memory and update size
+    size = totalNumBytes * sizeof(uint8_t) + nAnalogDataBytes +
+           nTransceiverDataBytes;
+
     memcpy(data + nAnalogDataBytes, result.data(),
            totalNumBytes * sizeof(uint8_t));
-    size = totalNumBytes * sizeof(uint8_t) + nAnalogDataBytes + ctbDbitOffset +
+
+    LOG(logDEBUG1) << "totalNumBytes: " << totalNumBytes
+                   << " nAnalogDataBytes:" << nAnalogDataBytes
+                   << " ctbDbitOffset:" << ctbDbitOffset
+                   << " nTransceiverDataBytes:" << nTransceiverDataBytes
+                   << " size:" << size;
+}
+
+/** ctb specific */
+void DataProcessor::ArrangeDbitData(size_t &size, char *data) {
```

Comment:
Can we note reuse this instead of "Reorder" and "Remove TRailing bits" ? Just need to check if ctb list is empty and not reorder, do nothing. Or am i missing something?